### DISCUSSION TITLE
Trapping emails sent inside event listeners with Japa

### DISCUSSION THREAD

**Discussion by khawarizmus** [adonisjs/core#3732] [Help] [answered]
My tests are failing as emails are not being sent within the event handler.
Given this code:
```ts
Event.on('user::created', async (user) => {
  // Emitted when a new user is created
  const token = await new TokenService().generateToken(user, TokenTypes.EMAIL)
  await new VerifyEmail(user, token).send()
})
```
Or this code (notice the removal of the async keyword)
```ts
Event.on('user::created', (user) => {
  // Emitted when a new user is created
  console.log('hello')
  console.log(user)
})
```
This test will always fail to catch the email and it also doesn't log anything from the event handler nor from the inside of `mailer.exists` function
```ts
test('Verify Email', async ({ client }, done) => {
    const mailer = Mail.fake()
    Event.emit('user::created', testUser)
      // check if email is sent
      await assert.isTrue(
        mailer.exists((mail) => {
          console.log(mail)
          done()
          return mail.subject === '[T9r9iba] Verify your email'
        })
      )
  }).waitForDone()
```
I have also tried using a `sleep` method before the `assert` call and it keeps failing. At this point i suspect that the event handler is not being executed at all and i am not sure if that's the actual behaviour or a bug.

Note: testing for the event capture works. for example the following test works fine
```ts
test('Register a new user (Signup)')
    .with([
      {
        username: 'test1234',
        email: 'test-register@email.com',
        password: 'test1234',
      },
    ])
    .run(async ({ client, assert }, userData) => {
      Mail.fake()
      const emitter = Event.fake()
      const response = await client.post('/auth/register').json(userData)
      const testUser = await User.findBy('email', userData.email)
      response.assertBodyContains({
        username: userData.username,
        email: userData.email,
        status: AccountStatus.PENDING,
      })
      // catch event
      assert.isTrue(
        emitter.exists((event) => {
          return event.name === 'user::created' && event.data.id === testUser?.id
        })
      )
      // check if user is created
      assert.isNotNull(testUser)
      // check if user is logged in (probably need to stub auth.login)
      const user = await User.findBy('id', response.session().auth_web)
      assert.isNotNull(user)
      assert.isTrue(user?.username === userData.username)
      // restore
      Event.restore()
      Mail.restore()
    })
```

I would appreciate if someone could guide me on this issue and explain how I can trap emails inside the event handlers.

**Comment by thetutlage** [comment, marked as answer]
Focusing on the following code block.  

```ts
test('Verify Email', async ({ client }, done) => {
    const mailer = Mail.fake()
    Event.emit('user::created', testUser)
      // check if email is sent
      await assert.isTrue(
        mailer.exists((mail) => {
          console.log(mail)
          done()
          return mail.subject === '[T9r9iba] Verify your email'
        })
      )
  }).waitForDone()

```

This is expected, because the events are emitted outside of the event loop. Infact, that is the purpose of events. Do not block the current event loop and do the thing in the background.

If you await the `Event.emit` call, then it will be part of the same event loop. Also, do not use `done` and `waitForDone`.

```ts
test('Verify Email', async ({ client }) => {
    const mailer = Mail.fake()
    await Event.emit('user::created', testUser)

    await assert.isTrue(
        mailer.exists((mail) => {
            console.log(mail)
            return mail.subject === '[T9r9iba] Verify your email'
        })
    )
  })
```

**Reply by khawarizmus** [reply]
@thetutlage That is still not working unfortunately. also my goal is to actually test the actual request which for simplicty i have only included the `Event.emit('user::created', testUser)` or event `await Event.emit('user::created', testUser)`.

but inside the controller i would probably not want to add an `await` keyword there.

**Reply by thetutlage** [reply]
Just tried. Works like a charm. Check this https://github.com/thetutlage/emitter-test

**Reply by khawarizmus** [reply]
@thetutlage can't see any event listeners in the project

**Reply by thetutlage** [reply]
https://github.com/thetutlage/emitter-test/blob/main/start/routes.ts#L25-L29

**Reply by khawarizmus** [reply]
This was solved by preloading the `events.js` file during testing environments. the way to do it was by adding the `test` entry in the `.adonisrc.json` file:
```json
"preloads": [
    "./start/routes",
    "./start/kernel",
    {
      "file": "./start/events",
      "environment": ["console", "repl", "web", "test"]
    }
  ],
  ```

**Comment by thetutlage** [comment]
Now, lemme share something which is a bit unrelated to the issue itself.

You are testing too many things in a single test. This is how it looks like to me right now.

- When I post data to `/auth/register`.
- A new user should be created.
- An event `user::created` should fire.
- That event should send an email to the user.

This needs to be split into multiple tests, as one test has too many responsibilities right now. It should be.

### First test

- When I post data to `/auth/register`.
- A new user should be created.
- An event `user::created` should fire.

```ts
.run(async ({ client, assert }, userData) => {
    const emitter = Event.fake()
    const response = await client.post('/auth/register').json(userData)

    response.assertStatus(200)
    response.assertBodyContains({
      username: userData.username,
      email: userData.email,
      status: AccountStatus.PENDING,
    })

    // catch event
    assert.isTrue(
      emitter.exists((event) => {
        return event.name === 'user::created' && event.data.id === testUser?.id
      })
    )

    // check if user is created
    const user = await User.findBy('id', response.session().auth_web)
    assert.isNotNull(user)
    assert.isTrue(user?.username === userData.username)

    // Restore emitter
    Event.restore()
  })
```

### Second test

It should test if the event is emitted, it does trigger the Mail.send function.

```ts
test('Verify Email', async ({ client }) => {
    const mailer = Mail.fake()
    await Event.emit('user::created', testUser)

    await assert.isTrue(
        mailer.exists((mail) => {
            console.log(mail)
            return mail.subject === '[T9r9iba] Verify your email'
        })
    )
  })
```

**Reply by khawarizmus** [reply]
Thank you for the advise. will keep that in mind moving forward.
